File: profiling/utils/uffd.py
import subprocess


class Uffd:
    def __init__(self) -> None:
        pass

    def get_uffd_pages(self) -> int:
        result = subprocess.run(f"tmux capture-pane -t uffd_handler -p", shell=True, capture_output=True, text=True)
        if result.returncode != 0:
            return -1
        lines = result.stdout.strip().split('\n')
        if lines:
            last_line = lines[-1]
            return int(last_line.split()[-1])

        else:
            raise ValueError("No output captured")

File: go_profiling/utils/uffd.go
package utils

import (
	"fmt"
	"os/exec"
	"strconv"
	"strings"
)

type Uffd struct{}

func (u *Uffd) GetUffdPages() (int, error) {
	cmd := exec.Command("tmux", "capture-pane", "-t", "uffd_handler", "-p")
	output, err := cmd.Output()
	if err != nil {
		return -1, err
	}
	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	if len(lines) > 0 {
		lastLine := lines[len(lines)-1]
		pageCount, err := strconv.Atoi(strings.Fields(lastLine)[len(strings.Fields(lastLine))-1])
		if err != nil {
			return -1, err
		}
		return pageCount, nil
	}
	return -1, fmt.Errorf("no output captured")
}
